docs/src/architecture/notes.md

```diff
@@ -1,73 +1,191 @@
+  
+      # compute the loop latch
+      dup dup.6 neq
+      # => [latch, ptr, 0, 0, 0, 0, end_ptr, ...]
+  
+      while.true
+          # => [ptr, 0, 0, 0, 0, end_ptr, ...]
+  
+          # save the pointer so that we can use it later
+          dup movdn.5
+          # => [ptr, 0, 0, 0, 0, ptr, end_ptr, ...]
+  
+          # load the asset and add it to the account
+          mem_loadw call.wallet::receive_asset
+          # => [ASSET, ptr, end_ptr, ...]
+  
+          # increment the pointer and compare it to the end_ptr
+          movup.4 add.1 dup dup.6 neq
+          # => [latch, ptr+1, ASSET, end_ptr, ...]
+      end
+  
+      # clear the stack
+      drop dropw drop
+  end
+  ```
+
+  The procedure starts by calling `exec.note::get_assets` and putting the note's number of assets and the memory pointer of the first asset on top of the stack. Assets are stored in consecutive memory slots, so `dup.1 add` provides the last memory slot. Because [assets](assets.md) are represented by `Words` in Miden Assembly, the procedure pads the stack with four `0`s. Now, if there is at least one asset (checked by `dup dup.6 neq`), the loop starts. It first saves the pointer for later use (`dup movdn.5`), then loads the first asset `mem_loadw` on top of the stack. Now, the procedure calls the a function of the account interface `call.wallet::receive_asset` to put the asset into the account's vault. The note script cannot directly call an account function to add the asset. The account must expose this function in its interface. Lastly, the pointer gets incremented, and if there is a second asset, the loop continues (`movup.4 add.1 dup dup.6 neq`). Finally, when all assets were put into the account's vault, the stack is cleared (`drop dropw drop`).
+
+</details>
+
+## Note storage mode
+Similar to accounts, there are two storage modes for notes in Miden. Notes can be stored publicly in the [Note DB](https://0xpolygonmiden.github.io/miden-base/architecture/state.html#notes-database) with all data publicly visible for everyone. Alternatively, notes can be stored privately by committing only the note hash to the Note DB. 
+
+Every note has a unique note hash. It is defined as follows `hash(hash(hash(hash(serial_num, [0; 4]), script_hash), input_hash), vault_hash)`. The note hash can be computed if all the note data is known. 
+
+_Info: To compute a note's hash, we do not need to know the note's `serial_num`. Knowing the hash of the `serial_num` (as well as `script_hash`, `input_hash` and `note_vault`) is also sufficient. We compute the hash of `serial_num` as `hash(serial_num, [0; 4])` to simplify processing within the VM._
+
+Privately stored notes can only be consumed if the note data is known to the consumer. The note data must be provided as input to the [transaction kernel](../transactions/transaction-kernel.md). That means, there must be some off-chain communication to transmit the note's data from the sender to the target.
```

Comment:
This is true for all notes (not just off-chain notes) - the consumer needs to know all note details.

Also, it feels like maybe this part belongs to the "Note consumption" section?